formula/src/main/java/com/instacart/formula/internal/ActionManager.kt

```diff
@@ -12,67 +14,157 @@ internal class ActionManager(
+
+    // SingleRequestMap for actions (lazy initialization)
+    private var actions: SingleRequestMap<Any, DeferredAction<*>>? = null
+
+    // For scheduling start/terminate operations in post-evaluation phase
+    private var checkToStartActionList: MutableList<DeferredAction<*>>? = null
+    private var checkToRemoveActionList: MutableList<DeferredAction<*>>? = null
+
     /**
-     * Currently running actions
+     * Enable or disable validation mode.
+     * Called by FormulaManagerImpl before each evaluation.
      */
-    private var running: LinkedHashSet<DeferredAction<*>>? = null
+    fun setValidationMode(enabled: Boolean) {
+        isValidationMode = enabled
+    }
 
     /**
-     * Action list provided by [Evaluation.actions]
+     * Find existing action by key or initialize new one.
+     * Schedules new actions for starting.
+     * Tracks keys for validation mode.
+     *
+     * IMPORTANT: Unlike listeners/children, actions do NOT support indexed collision handling.
+     * If a key collision occurs, it's treated as an error.
      */
-    private var actions: Set<DeferredAction<*>> = emptySet()
+    fun <Event> findOrInitAction(
+        key: Any,
+        action: Action<Event>,
+        listener: Listener<Event>
+    ): DeferredAction<Event> {
+        // Lazy init actions map
+        val actionsMap = actions ?: run {
+            val map = mutableMapOf<Any, SingleRequestHolder<DeferredAction<*>>>()
+            actions = map
+            map
+        }
 
-    private var recomputeCheckToStartList: Boolean = false
-    private var checkToStartActionList: MutableList<DeferredAction<*>>? = null
+        // Check if this is a new action (key doesn't exist yet)
+        val isNew = !actionsMap.containsKey(key)
 
-    private var recomputeCheckToRemoveList: Boolean = false
-    private var checkToRemoveActionList: MutableList<DeferredAction<*>>? = null
+        val holder = actionsMap.findOrInit(key) {
+            DeferredAction(key, action, listener)
+        }
+
+        if (holder.requested) {
+            // Key collision - this is an error
+            val formulaType = manager.formulaType
+            val message = "$formulaType - duplicate action key in same evaluation: $key. " +
+                    "This likely means the same action is being registered multiple times. " +
+                    "Ensure each action has a unique key or is only registered once per evaluation."
+            throw IllegalStateException(message)
+        }
+
+        // Mark as requested for this evaluation cycle
+        holder.requested = true
+
+        // Schedule for starting if it's a new action
+        if (isNew) {
+            if (isValidationMode) {
+                val errors = validationErrors ?: LinkedList<ActionValidationError>().also {
+                    validationErrors = it
+                }
+                errors.add(ActionValidationError.NewAction(key))
```

Comment:
Not sure I'm following this - why is this error added here?

Reply:
Simplified validation errors, hopefully much easier to reason about.